packages/core/contracts/oracle/implementation/StakerSnapshot.sol

```diff
@@ -0,0 +1,242 @@
+
+    // If: a staker requested an unstake and time > unstakeTime then send funds to staker. Note that this method assumes
+    // that the `updateTrackers()
+    function executeUnstake() public {
+        _updateTrackers(msg.sender);
+        updateSnapshot(msg.sender);
+        VoterStake storage voterStake = voterStakes[msg.sender];
+        require(voterStake.unstakeTime != 0 && getCurrentTime() >= voterStake.unstakeTime, "Unstake time not passed");
+        uint256 tokensToSend = voterStake.requestUnstake;
+        if (tokensToSend > voterStake.cumulativeStaked) tokensToSend = voterStake.cumulativeStaked;
+
+        if (tokensToSend > 0) {
+            voterStake.cumulativeStaked -= tokensToSend;
+            cumulativeStaked -= voterStake.requestUnstake;
+            voterStake.requestUnstake = 0;
+            voterStake.unstakeTime = 0;
+            votingToken.transfer(msg.sender, tokensToSend);
+        }
+    }
+
+    // Send accumulated rewards to the voter. If the voter has gained rewards from others slashing then this is included
+    // here. If the total slashing is larger than the outstanding rewards then this method does nothing.
+    function withdrawRewards() public {
+        _updateTrackers(msg.sender);
+        VoterStake storage voterStake = voterStakes[msg.sender];
+
+        if (voterStake.outstandingRewards > 0) {
+            require(votingToken.mint(msg.sender, voterStake.outstandingRewards), "Voting token issuance failed");
+            voterStake.outstandingRewards = 0;
+        }
+    }
+
+    function exit() public {
+        executeUnstake();
+        withdrawRewards();
+    }
+
+    function _updateTrackers(address voterAddress) internal {
+        _updateReward(voterAddress);
+    }
+
+    // Calculate the reward per token based on last time the reward was updated.
+    function _updateReward(address voterAddress) internal {
+        rewardPerTokenStored = rewardPerToken();
+        lastUpdateTime = getCurrentTime();
+        if (voterAddress != address(0)) {
+            VoterStake storage voterStake = voterStakes[voterAddress];
+            voterStake.outstandingRewards = outstandingRewards(voterAddress);
+            voterStake.rewardsPaidPerToken = rewardPerTokenStored;
+        }
+    }
+
+    function outstandingRewards(address voterAddress) public view returns (uint256) {
+        VoterStake storage voterStake = voterStakes[voterAddress];
+
+        return
+            ((voterStake.cumulativeStaked * (rewardPerToken() - voterStake.rewardsPaidPerToken)) / 1e18) +
+            voterStake.outstandingRewards;
+    }
+
+    function rewardPerToken() public view returns (uint256) {
+        if (cumulativeStaked == 0) return rewardPerTokenStored;
+        return rewardPerTokenStored + ((getCurrentTime() - lastUpdateTime) * emissionRate * 1e18) / cumulativeStaked;
+    }
+
+    // Owner methods
+    function setEmissionRate(uint256 _emissionRate) public onlyOwner {
+        emissionRate = _emissionRate;
```

Comment:
I believe that currently, if the emissionRate is changed and there are unclaimed rewards they will be retroactively updated.  Is this a problem?

I have been exporting this idea in this branch, to better understand the mechanism. (Maybe it is not relevant at all) : https://github.com/UMAprotocol/protocol/pull/3987

Reply:
this is an excellent catch!